internals/overlord/servstate/workloads.go

```diff
@@ -0,0 +1,209 @@
+// Copyright (c) 2025 Canonical Ltd
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package servstate
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+
+	yaml "gopkg.in/yaml.v3"
+
+	"github.com/canonical/pebble/internals/plan"
+)
+
+var _ plan.SectionExtension = (*WorkloadsSectionExtension)(nil)
+
+type WorkloadsSectionExtension struct{}
+
+func (ext *WorkloadsSectionExtension) CombineSections(sections ...plan.Section) (plan.Section, error) {
+	ws := &WorkloadsSection{}
+	for _, section := range sections {
+		layer, ok := section.(*WorkloadsSection)
+		if !ok {
+			return nil, fmt.Errorf("internal error: invalid section type %T", layer)
+		}
+		if err := ws.combine(layer); err != nil {
+			return nil, err
+		}
+	}
+	return ws, nil
+}
+
+func (ext *WorkloadsSectionExtension) ParseSection(data yaml.Node) (plan.Section, error) {
+	ws := &WorkloadsSection{}
+	// The following issue prevents us from using the yaml.Node decoder
+	// with KnownFields = true behavior. Once one of the proposals get
+	// merged, we can remove the intermediate Marshal step.
+	if len(data.Content) != 0 {
+		yml, err := yaml.Marshal(data)
+		if err != nil {
+			return nil, fmt.Errorf(`internal error: cannot marshal "workloads" section: %w`, err)
+		}
+		dec := yaml.NewDecoder(bytes.NewReader(yml))
+		dec.KnownFields(true)
+		if err = dec.Decode(ws); err != nil {
+			return nil, &plan.FormatError{
+				Message: fmt.Sprintf(`cannot parse the "workloads" section: %v`, err),
+			}
+		}
+	}
+	for name, workload := range ws.Entries {
+		if workload != nil {
+			workload.Name = name
+		}
+	}
+	return ws, nil
+}
+
+func (ext *WorkloadsSectionExtension) ValidatePlan(p *plan.Plan) error {
+	ws, ok := p.Sections[WorkloadsField].(*WorkloadsSection)
+	if !ok {
+		return fmt.Errorf("internal error: invalid section type %T", ws)
+	}
+	for name, service := range p.Services {
+		_, ok := ws.Entries[service.Workload]
+		if service.Workload != "" && !ok {
+			return &plan.FormatError{
+				Message: fmt.Sprintf(`plan service %q cannot run in unknown workload %q`, name, service.Workload),
+			}
+		}
+	}
+	return nil
+}
+
+const WorkloadsField = "workloads"
+
+var _ plan.Section = (*WorkloadsSection)(nil)
+
+type WorkloadsSection struct {
+	Entries map[string]*Workload `yaml:",inline"`
+}
+
+func (ws *WorkloadsSection) IsZero() bool {
```

Comment:
Suggestion to rename to `IsEmpty`

Reply:
Feels more fitting, however `IsZero()` is already part of the `Section` interface in the plan:

https://github.com/canonical/pebble/blob/d8a630a02868fa5347bae81494cc17007b4ef231/internals/plan/plan.go#L53-L59

Reply:
Which comes from https://pkg.go.dev/gopkg.in/yaml.v3#IsZeroer :)